database.ml

```diff
@@ -65,66 +65,197 @@ let execute_insert_returning (table_name, field_names, vss, returning) db =
                              (fun row ->
                                `Record (
-                                 List.fold_right
+                                List.fold_right
                                    (fun (name, (t, i)) fields ->
                                      (name, value_of_db_string (List.nth row i) t)::fields)
                                    fields
-                                   []))
-                      result#get_all_lst)
+                                   []
+			       ))
+                      	(Debug.debug_time "get_all_lst" (fun () -> result#get_all_lst))
+		      )
+)
+
+*)
+
+(* BUG: Lists can be too big for List.map; need to be careful about recursion *)
+
+let execute_select_result
+    (field_types:(string * Types.datatype) list) (query:string) (db: database)  =
+  let result = (db#exec query) in
+    (match result#status with
+       | `QueryOk -> 
+           result,
+	   result_signature field_types result
+       | `QueryError msg -> raise (Runtime_error ("An error occurred executing the query " ^ query ^ ": " ^ msg)))
+
+(* Experimental code to build result using getvalue *)
+let build_result_deforest ((result:Value.dbvalue),rs) = 
+  let max = result#ntuples in
+  let rec do_map n acc = 
+    let rec build rs l = 
+      match rs with
+      | [] -> l
+      | (name,(t,i))::rs' -> 
+	  build rs' ((name,value_of_db_string (result#getvalue n i) t)::l)
+      | _ -> assert false in
+    if n < max
+    then do_map (n+1) (`Record(build rs [])::acc)
+    else `List (acc)
+  in do_map 0 []  
+    ;;
+    
+let build_result_fun ((result:Value.dbvalue),rs) = 
+  `List (result#map (fun row ->
+                     `Record (build_record rs row))
+	   )
+    ;;
+
+(* experimental *)
+let build_result_array ((result:Value.dbvalue),rs) = 
+  `List (result#map_array (fun row ->
+                     `Record (build_record_array rs row))
+	   )
+    ;;
+
+let build_result (result,rs) = build_result_fun (result,rs)
+;;
+
+let execute_select_old
+    (field_types:(string * Types.datatype) list) (query:string) (db : database)
+    : Value.t =
+  let result = db#exec query in
+  
+  match result#status with
+  | `QueryError msg -> 
+      raise(Runtime_error("An error occurred executing the query " ^ query ^
+                          ": " ^ msg))
+  | `QueryOk -> 
+      Debug.debug_time "QueryOK" (fun () -> 
+	let rs = result_signature field_types result in
+	`List (result#map
+                 (fun row ->
+                   `Record (build_record rs row)
+		     )
+		 )
+	  )
+	
+    
+	
+
+
+let execute_select_new
+    (field_types:(string * Types.datatype) list) (query:string) (db : database)
+    : Value.t =
+  let result,rs = execute_select_result field_types query db in 
+  build_result (result,rs)
+;;
+    
+
+let execute_select  
+  (field_types:(string * Types.datatype) list) (query:string) (db : database)
+    : Value.t =
+  if Settings.get_value Basicsettings.fast_execute_select 
```

Comment:
We can probably hardwire this to the "true" branch, since there is nothing shredding-specific about this